### ISSUE TITLE
Auto-Bootstrap Procedure

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#2020]
As described in the manual, Spack works best (espcially on older machiens) if you use a number of Spack-built packages in the environment used to run Spack.  This creates a somewhat tedious bootstrapping procdure, in which the Spack admin has to check around the system for what is out of date, and then use Spack to install newer versions of those things.  And you have to do it in the right order too (for example, some Spack packages won't download if you have SSL problems).

Now we have Spack depending on a YAML parser as well (for best performance).

Maybe it's time to build an auto-bootstrap procedure.  When you give Spack the 'spack auto-bootstrap` command, it would root around your system, figure out what needs to be built, build it, and then include those packages in the Spack path forever after (even if you never put them in your .bashrc).

I think this could make it MUCH easier to install Spack on a wide variety of systems.  Thoughts?

**Comment by tgamblin** [comment]
I don't think that there should be a `spack auto-bootstrap` command -- if Spack fails to find a system `curl` (or other packages) on startup it should just install them and the user shouldn't have to run a command.

I only think that things *necessary* to run (e.g., `curl`) should be installed like this.  I don't think YAML fits that bill -- we ship with Python YAML and it works fine.  That's an optimization, NOT a dependency.  And it is not something anyone needs immediately -- only after several hundred packages.

Finally, this shouldn't be a special procedure.  We currently have custom logic to detect compilers by searching around the `PATH` and modules.  This should be gentrified to find compilers and dependency packages and add them to `packages.yaml`, and that is how Spack should know what is installed -- using the existing database.

@mamelara is working on a `spack external add` feature that implements some of this -- it's designed so that you can point at an external package and have Spack automatically put them in the `packages.yaml` file.  I expect that when compiler become dependencies they will move there as well.

**Comment by citibeth** [comment]
On Fri, Oct 14, 2016 at 11:24 AM, Todd Gamblin <notifications@github.com>
wrote:

> I don't think that there should be a spack auto-bootstrap command -- if
> Spack fails to find a system curl (or other packages) on startup it
> should just install them and the user shouldn't have to run a command.
>
I like this idea.

> I only think that things *necessary* to run (e.g., curl) should be
> installed like this. I don't think YAML fits that bill -- we ship with
> Python YAML and it works fine. That's an optimization, NOT a dependency.
> And it is not something anyone needs immediately -- only after several
> hundred packages.
>
This is a matter of degree.  Spack works find out-of-the-box.  Except that
if your SSL is bad, you'll have trouble as soon as you try to install GCC.
But you can install packages NOT requiring SSL (eg, OpenSSL) just fine.

And... if Spack picks up things it NEEDS with as little hassel as possible,
it should be able to do the same thing later on for optimizations.  I
agree, it shouldn't be done automatically.  Maybe we need a `spack
optimize` command that does all this.


> Finally, this shouldn't be a special procedure. We currently have custom
> logic to detect compilers by searching around the PATH and modules. This
> should be gentrified to find compilers and dependency packages and add them
> to packages.yaml, and that is how Spack should know what is installed --
> using the existing database.
>

Not sure what you mean by this.  The proposed `auto-bootstrap` command
would run the required `spack install`s.

**Comment by tgamblin** [comment]
I think we're on the same page I just don't want people to have to run `auto-bootstrap`.

And the API should be thought out.  Right now there are a lot of places where Spack just calls `which`, and those should be replaced with some wrapper around the DB that installs a thing if it's needed and returns an `Executable`.  Likewise, there should be a similar install-if-needed for libraries.  And this should do something graceful if it's not connected to the internet.

Basically I am just saying that reworking this properly is a midsize project, and we should do it eventually, but we should think it through.  We want to minimize the number of dependencies Spack has, and we don't want this in the mainline until it is really stable, or there are going to be bugs that keep people from using Spack.

**Comment by adamjstewart** [comment]
> Finally, this shouldn't be a special procedure. We currently have custom logic to detect compilers by searching around the PATH and modules. This should be generified to find compilers and dependency packages and add them to packages.yaml, and that is how Spack should know what is installed -- using the existing database.

> @mamelara is working on a spack external add feature that implements some of this -- it's designed so that you can point at an external package and have Spack automatically put them in the packages.yaml file. I expect that when compiler become dependencies they will move there as well.

This sounds awesome! I was wondering how we were going to address the growing complexity problem. The problem is that packages may depend on common system libraries like Xorg/X11, curl, openssl, etc., but these libraries may not exist on _every_ system. These dependencies should be packaged into Spack so that Spack doesn't rely on the system package manager, but there should also be an easy option to use system libraries. It sounds like this functionality is an automatic way to scrape the system for all available dependencies and be able to use them within Spack to satisfy dependencies. I think this is a great option going forward and users can easily include or exclude packages that they do or don't care about building with Spack. This way I don't have to rebuild a million dependencies every time the hash changes or a new version of zlib is released. But I can rebuild things every time a new version of hdf5 comes out which I care about.

**Comment by citibeth** [comment]
Yes!  It would be great if the feature can root around your (standardly
installed) system and find any Spackable packages you might already have.

**Comment by tgamblin** [comment]
We've talked about this a bit with OpenHPC folks, as well, and there was some motivation to have system packages provide version information in a standard, Spack-spec-ish way.  Basically something to get you a homebrew-like experience on a non-mac system.  Homebrew "just knows" what is on your system, and systems that use RPM, etc. can tell you what's there, but the provenance information they provide is in lots of different formats.  We either need something to, say, scrape the RPM DB for known RHEL/Fedora/SuSE package names (sounds unreliable to me) or we could potentially get the distro people to adopt a provenance format like `spec.yaml`.  OpenHPC seems willing to do something like the latter.

**Comment by citibeth** [comment]
I think this is a little beyond my original post.  My ORIGINAL idea was just to automate the procedure that is currently described manually in the "Getting Started" section of the manual.

**Comment by alalazo** [comment]
Is it fine to close this now that #3057 got in?
